File: renderers/qwen36.py
"""Qwen3.6 Renderer — mirrors the Qwen3.6 Jinja chat template.

Delta vs Qwen3.5 (template line 122):

- Tool-call argument serialization changed to ``tojson`` for every non-string
  value. Bools now render as ``true``/``false`` (not ``True``/``False``) and
  ``None`` as ``null`` (not ``None``), fixing the single-turn extension-break
  mode where a boolean parameter's case drifted across a re-render.

Historical-thinking retention follows Qwen3.5's default (drop past
``<think>`` blocks). The upstream template carries a ``preserve_thinking``
Jinja toggle for the opposite polarity; on the renderer side that intent
maps to the renderer-agnostic ``preserve_all_thinking`` /
``preserve_thinking_between_tool_calls`` flags on
:class:`renderers.Qwen36RendererConfig`.

Everything else — tool system prompt, tool-call XML structure, thinking
markers, bridge logic, parser — is identical to Qwen3.5.
"""

from __future__ import annotations

import json
from typing import Any

from renderers._native_router import (
    load_native,
    native_enabled,
    resolve_tokenizer_path,
)
from renderers.configs import Qwen36RendererConfig
from renderers.qwen35 import Qwen35Renderer, _default_enable_thinking


class Qwen36Renderer(Qwen35Renderer):
    """Deterministic message → token renderer for Qwen3.6 models."""

    _config_cls = Qwen36RendererConfig

    def __new__(
        cls,
        tokenizer,
        config: Qwen36RendererConfig | None = None,
        *,
        processor=None,
    ):
        # Route to native only for Qwen3.6 specifically — never fall
        # through to the parent's qwen35 router (the renderer flag is
        # different).
        if native_enabled("qwen36") and processor is None:
            native = load_native()
            if native is not None:
                cfg = config or Qwen36RendererConfig()
                enable_thinking = cfg.enable_thinking
                if enable_thinking is None:
                    enable_thinking = _default_enable_thinking(tokenizer)
                path = resolve_tokenizer_path(tokenizer)
                return native.Renderer.qwen36(
                    path,
                    enable_thinking=enable_thinking,
                    preserve_all_thinking=cfg.preserve_all_thinking,
                    preserve_thinking_between_tool_calls=cfg.preserve_thinking_between_tool_calls,
                )
        # Skip Qwen35Renderer.__new__ (would also try to route, with the
        # wrong flag). Go straight to object.
        return object.__new__(cls)

    @staticmethod
    def _render_arg_value(arg_value: Any) -> str:
        if isinstance(arg_value, str):
            return arg_value
        return json.dumps(arg_value, ensure_ascii=False)

File: crates/renderers-core/src/families/qwen36.rs
//! Qwen3.6 renderer. Delta vs Qwen3.5: tool-call arguments serialise
//! through JSON (bools → `true`/`false`, None → `null`, etc.) instead of
//! Python `str()`. Everything else — template structure, parser,
//! tool-call XML, thinking markers, bridge logic — is identical to
//! Qwen3.5, so this is a one-line config delta on the Qwen3.5 builder.
//!
//! Mirrors `renderers/qwen36.py`.

use crate::families::Qwen35RendererBuilder;

/// Build a Qwen3.6 renderer.
///
/// Type alias preserved as a re-export of [`Qwen35Renderer`](crate::families::Qwen35Renderer)
/// — the type system doesn't distinguish them at runtime; they differ
/// only in the `args_as_json` flag. The builder below is the right
/// public surface.
pub use crate::families::Qwen35Renderer as Qwen36Renderer;

/// Builder for [`Qwen36Renderer`] (a Qwen3.5 with JSON-flavoured tool
/// arguments).
#[derive(Debug, Clone, Default)]
pub struct Qwen36RendererBuilder {
    inner: Qwen35RendererBuilder,
}

impl Qwen36RendererBuilder {
    pub fn enable_thinking(mut self, on: bool) -> Self {
        self.inner = self.inner.enable_thinking(on);
        self
    }
    pub fn preserve_all_thinking(mut self, on: bool) -> Self {
        self.inner = self.inner.preserve_all_thinking(on);
        self
    }
    pub fn preserve_thinking_between_tool_calls(mut self, on: bool) -> Self {
        self.inner = self.inner.preserve_thinking_between_tool_calls(on);
        self
    }
    pub fn build(
        self,
        tokenizer: crate::tokenizer::Tokenizer,
    ) -> Result<Qwen36Renderer, crate::types::RenderError> {
        self.inner.args_as_json(true).build(tokenizer)
    }
}
